py/server/deephaven/experimental/iceberg.py

```diff
@@ -0,0 +1,251 @@
+                 data_instructions: Optional[s3.S3Instructions] = None,
+                 column_renames: Optional[Dict[str, str]] = None):
+        """
+        Initializes the instructions using the provided parameters.
+
+        Args:
+            table_definition (Optional[Union[Dict[str, DType], List[Column], None]]): the table definition; if omitted,
+                the definition is inferred from the Iceberg schema. Setting a definition guarantees the returned table
+                will have that definition. This is useful for specifying a subset of the Iceberg schema columns.
+            data_instructions (Optional[s3.S3Instructions]): Special instructions for reading data files, useful when
+                reading files from a non-local file system, like S3.
+            column_renames (Optional[Dict[str, str]]): A dictionary of old to new column names that will be renamed in
+                the output table.
+
+        Raises:
+            DHError: If unable to build the instructions object.
+        """
+
+        try:
+            builder = self.j_object_type.builder()
+
+            if table_definition is not None:
+                builder.tableDefinition(j_table_definition(table_definition))
+
+            if data_instructions is not None:
+                builder.dataInstructions(data_instructions.j_object)
+
+            if column_renames is not None:
+                for old_name, new_name in column_renames.items():
+                    builder.putColumnRenames(old_name, new_name)
+
+            self._j_object = builder.build()
+        except Exception as e:
+            raise DHError(e, "Failed to build Iceberg instructions") from e
+
+    @property
+    def j_object(self) -> jpy.JType:
+        return self._j_object
+
+
+class IcebergCatalogAdapter(JObjectWrapper):
+    """
+    This class provides an interface for interacting with Iceberg catalogs. It allows listing namespaces, tables and
+    snapshots, as well as reading Iceberg tables into Deephaven tables.
+    """
+    j_object_type = _JIcebergCatalogAdapter or type(None)
+
+    def __init__(self, j_object: _JIcebergCatalogAdapter):
+        self.j_catalog_adapter = j_object
+
+    def namespaces(self, namespace: Optional[str] = None) -> Table:
+        """
+        Returns the namespaces in the catalog as a Deephaven table.
+
+        Args:
+            namespace (Optional[str]): the higher-level namespace from which to list namespaces; if omitted, the
+                top-level namespaces are listed.
+
+        Returns:
+            a table containing the namespaces.
+        """
+
+        if namespace is not None:
+            return Table(self.j_object.listNamespaces(namespace))
+        return Table(self.j_object.listNamespacesAsTable())
+
+    def tables(self, namespace: str) -> Table:
+        """
+        Returns the list of tables in the provided namespace as a Deephaven table.
```

Comment:
1. Is this list of tables just a single column?  If not, there should be some indication.
2. Is the list of tables the `table_identifier`s listed in `snapshots` and possibly other methods.  This is not clear, so I'm left guessing that they are they same.

Reply:
Exploration of the namespaces and tables is provided, but we expect the typical user of Iceberg to already have a firm grasp of the structure of their data. 

For completeness, here are what the tables look like when the functions are called:

`namespaces()`:
<img width="513" alt="image" src="https://github.com/deephaven/deephaven-core/assets/4204632/87785bbd-53e9-4e4f-a92c-4fa1e4f0e841">

`tables()`:
<img width="513" alt="image" src="https://github.com/deephaven/deephaven-core/assets/4204632/b833ff97-4153-4a6c-94ea-eeac5c5a9b62">

`snapshots()`:
<img width="2016" alt="image" src="https://github.com/deephaven/deephaven-core/assets/4204632/6631bcf4-1b9f-4189-8fc1-6efa41953a49">

Reply:
One thing I do notice is that the columns are snake case, when I think DH has tried to use camel case for everything.

Reply:
Ryan also pointed this out. The Iceberg convention is snake case inside its string descriptions and internal maps and I followed its convention. I'm ambivalent though.

Reply:
I would do DH convention, unless there is a compelling reason not to.